File: javascript/lasagna/lasagna.js
// @ts-check

/**
 * The number of minutes it takes to prepare a single layer.
 */
const PREPARATION_MINUTES_PER_LAYER = 2;

/**
 * The expected oven time in minutes.
 */
export const EXPECTED_MINUTES_IN_OVEN = 40;

/**
 * Determines the number of minutes the lasagna still needs to remain in the
 * oven to be properly prepared.
 *
 * @param {number} actualMinutesInOven already elapsed baking time
 * @returns {number} the number of minutes remaining
 */
export function remainingMinutesInOven(actualMinutesInOven) {
  return EXPECTED_MINUTES_IN_OVEN - actualMinutesInOven;
}

/**
 * Given a number of layers, determines the total preparation time.
 *
 * @param {number} numberOfLayers the number of layers in the lasagna
 * @returns {number} the total preparation time
 */
export function preparationTimeInMinutes(numberOfLayers) {
  return PREPARATION_MINUTES_PER_LAYER * numberOfLayers;
}

/**
 * Calculates total working time. That is, the time to prepare all the layers
 * of lasagna, and the time already spent in the oven.
 *
 * @param {number} numberOfLayers the number of layers in the lasagna
 * @param {number} actualMinutesInOven elapsed cooking time
 * @returns {number} total working elapsed time in preparing and cooking
 */
export function totalTimeInMinutes(numberOfLayers, actualMinutesInOven) {
  return preparationTimeInMinutes(numberOfLayers) + actualMinutesInOven;
}

File: go/lasagna/lasagna.go
package lasagna

// The expected oven time in minutes & minutes to prepare a single layer.
const OvenTime, minutesPerLayer = 40, 2

// RemainingOvenTime returns the remaining minutes based on the `actual`
// minutes already in the oven.
func RemainingOvenTime(elapsed int) int { return OvenTime - elapsed }

// PreparationTime calculates the time needed to prepare the lasagna based on
// the amount of layers.
func PreparationTime(layers int) int { return layers * minutesPerLayer }

// ElapsedTime calculates the time elapsed cooking the lasagna. This time
// includes the preparation time and the time the lasagna is baking in the oven.
func ElapsedTime(lays, time int) int { return PreparationTime(lays) + time }
